.github/workflows/ci2.yaml

```diff
@@ -0,0 +1,360 @@
+    needs: [build]
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+          - package: plutus-cbor
+          - package: plutus-merkle-tree
+          - package: hydra-plutus
+          - package: hydra-tui
+          - package: hydra-node
+          - package: hydra-cluster
+    steps:
+    - name: 📥 Checkout repository
+      uses: actions/checkout@v3
+      with:
+        repository: input-output-hk/hydra
+        token: ${{ secrets.MY_TOKEN || github.token }}
+        # On pull_request events, we want to check out the latest commit of the
+        # PR, which is different to github.ref (the default, which would point
+        # to a "fake merge" commit). On push events, the default is fine as it
+        # refers to the pushed commit.
+        ref: ${{ github.event.pull_request.head.sha || github.ref }}
+        # Also ensure we have all history with all tags
+        fetch-depth: 0
+
+    - uses: ./.github/workflows/actions/setup_dev_ubuntu
+      with:
+        clean_caches: false
+    
+    - name: ❓ Test
+      if: ${{ matrix.package != 'hydra-tui' }}
+      run: |
+        cabal test ${{ matrix.package }}
+
+    - name: ❓ Test (TUI)
+      id: test_tui
+      if: ${{ matrix.package == 'hydra-tui' }}
+      # https://giters.com/gfx/example-github-actions-with-tty
+      # The default shell does not allocate a TTY which breaks some tests
+      shell: 'script -q -e -c "bash {0}"'
+      env:
+        TERM: "xterm"
+      # TUI specs are flaky. They are failing because of SIGSEGV.
+      # There is an open issue to tackle this problem. https://github.com/input-output-hk/hydra/issues/590
+      continue-on-error: true
+      run: |
+        cabal test ${{ matrix.package }}
+
+    - name: 💾 Upload build & test artifacts
+      uses: actions/upload-artifact@v3
+      with:
+        name: test-results
+        path: |
+          ./**/test-results.xml
+          ./**/hspec-results.md
+
+
+    # NOTE: This depends on the path used in hydra-cluster e2e tests
+    - name: 💾 Upload logs
+      if: always()
+      uses: actions/upload-artifact@v3
+      with:
+        name: hydra-cluster-e2e-test-logs
+        path: /tmp/hydra-cluster-e2e-*/logs/*
+
+  publish-test-results:
+    name: Publish test results
+    needs: [test]
+    runs-on: ubuntu-latest
+    steps:
+    - name: 📥 Download test results
+      uses: actions/download-artifact@v3
+      with:
+        name: test-results
+
+    - name: ✏ Publish test results to PR
+      uses: EnricoMi/publish-unit-test-result-action@v2
+      with:
+        junit_files: ./**/test-results.xml
+
+  haddock:
+    name: "Haddock"
+    # needs: [build]
+    runs-on: ubuntu-latest
+    steps:
+    - name: 📥 Checkout repository
+      uses: actions/checkout@v3
+      with:
+        repository: input-output-hk/hydra
+        token: ${{ secrets.MY_TOKEN || github.token }}
+
+    - uses: ./.github/workflows/actions/setup_dev_ubuntu
+      with:
+        cache_prefix: 'haddock'
```

Comment:
Good question. It has not been super intuitive at first, for me, to grasp the logic behind [GitHub caches](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows) but it's not that bad actually.

Long story short:
* _build_ and _haddock_ generate very different compilation outputs I'm not sure exactly how it works but haddock is taking forever if only starting from the _build_ cache.
* if we set _haddock_ to use the same cache as _build_ it will not update this cache to add stuff in it, a cache is immutable so it means that, next time we run haddock, it will, again, rebuild everything that's missing in the _build_ cache
* we have the same issue if we do the opposite, using the _haddock_ cache for the _build_ step. Although I did not test, I would expect _build_ to need to build stuff that _haddock_ does not need
* if we run both job in parallel on the same cache, the first one to finish will generate the cache (or the last one, not sure) and we're in a similar situation with one deciding what's in the cache for the other.

Reply:
OK... now that the big mystery has been [solved](https://github.com/input-output-hk/hydra/wiki/Logbook-2023-H1#pgab-on-ci-experimentation), it's using the same cache as the rest 🎉